modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsFile.java

```diff
@@ -26,23 +26,57 @@
 import org.apache.hadoop.fs.FSDataOutputStream;
 import org.apache.hadoop.fs.Path;
 
+import java.io.IOException;
+
 public class HdfsFile {
     private FSDataOutputStream fsDataOutputStream;
     private Path path;
+    private long lastWrite;
 
-    public FSDataOutputStream getFsDataOutputStream() {
-        return fsDataOutputStream;
-    }
-
-    public void setFsDataOutputStream(FSDataOutputStream fsDataOutputStream) {
-        this.fsDataOutputStream = fsDataOutputStream;
+    public HdfsFile(Path filepath, FSDataOutputStream outputstream) {
+        this.path = filepath;
+        this.fsDataOutputStream = outputstream;
+        this.lastWrite = 0;
     }
 
     public Path getPath() {
         return path;
     }
 
-    public void setPath(Path path) {
-        this.path = path;
+    public boolean isOpen() {
+        return fsDataOutputStream != null;
+    }
+
+    public void flush() throws IOException {
+         if (!isOpen())
+            return;
+
+         fsDataOutputStream.hflush();
+         fsDataOutputStream.hsync();
+    }
+
+    public void write(byte[] message) throws IOException {
+         if (!isOpen())
+            throw new IOException("File is not open: "+path);
+
+         fsDataOutputStream.write(message);
+
+         lastWrite = System.currentTimeMillis();
+    }
+
+    public long timeSinceLastWrite() {
+       if (!isOpen())
```

Comment:
I think this shouldn't be here, it should never happen that we call this function and the file is not open.
currently we just hide if there is something wrong with the code outside

Reply:
Well yes, it is not called now. But that does not mean it should not be here at all. As the `HdfsFile` is a separate class that should work regardless how it is called.
This class has an internal state of not open, that should be respected by all of its method.

Reply:
can you tell me the decision logic when to raise an exception and when to do other things if the file is not open?
(e.g.: write will thrown an exception, flush will does not nothing, this will return with zero)

Reply:
* `flush`: there was no decision on my part at all it should be identical to the code before.
* `write`: here the logic was to keep the stream write behavior as throwing an exception
* `timeSinceLastWrite`: the `return 0;` just meant as `return`, but I cannot do the later one in case of return value